### ISSUE TITLE
[Bug]: plt.hist takes significantly more time with torch and jax arrays

### ISSUE THREAD

**Issue by patel-zeel** [matplotlib/matplotlib#25882]
### Bug summary

Hi,

Time taken to plot `plt.hist` directly on `jax` or `torch` arrays is significantly more than **combined time taken to first convert them to `numpy` and then using `plt.hist`**. Shouldn't `matplotlib` internally convert them to `numpy` arrays before plotting?

To reproduce the bug, directly run the following snippet on Google Colab.

### Code for reproduction

```python
from time import time
import numpy as np

import torch

import jax
import jax.random as jr
import jax.numpy as jnp

import matplotlib.pyplot as plt

jax_array = jr.normal(jr.PRNGKey(0), (1000, 150))
torch_array = torch.randn(1000, 150)

def plot_hist(array):
    init = time()
    plt.figure()
    plt.hist(array)
    print(f"Time to plot: {time() - init:.2f} s")
    plt.show()
    
plot_hist(jax_array.ravel())
plot_hist(torch_array.ravel())
plot_hist(np.array(jax_array.ravel()))
plot_hist(np.array(torch_array.ravel()))
```

### Actual outcome

Time to plot: 4.19 s
![image](https://github.com/matplotlib/matplotlib/assets/59758528/29f7e6e6-4c91-4801-8472-24eaaae591cc)

Time to plot: 2.61 s
![image](https://github.com/matplotlib/matplotlib/assets/59758528/cf0b12c2-fabf-45fb-8e59-89206bb6ed68)

Time to plot: 0.03 s
![image](https://github.com/matplotlib/matplotlib/assets/59758528/9a39702e-b0ba-48ba-9ec6-5731f3d4b372)

Time to plot: 0.04 s
![image](https://github.com/matplotlib/matplotlib/assets/59758528/5e99fa5c-c614-4422-9b98-ff62fa3d3563)

### Expected outcome

Time to plot: 0.03 s

Time to plot: 0.04 s

Time to plot: 0.03 s

Time to plot: 0.04 s

### Additional information

> What are the conditions under which this bug happens? input parameters, edge cases, etc?

It is happening with all kinds of shapes.

> Has this worked in earlier versions?

Tested with default colab `matplotlib` version 3.7.1 and also with 3.6.3.

> Do you know why this bug is happening?

Not exactly sure.

> Do you maybe even know a fix?

Maybe convert any python object to a `numpy` array before plotting?

### Operating system

Ubuntu 20.04.5 LTS

### Matplotlib Version

3.7.1

### Matplotlib Backend

module://matplotlib_inline.backend_inline

### Python version

3.10.11

### Jupyter version

6.4.8

### Installation

None

**Comment by oscargus** [comment]
The unpacking happens here:
https://github.com/matplotlib/matplotlib/blob/b61bb0b6392c23d38cd45c658bfcd44df145830d/lib/matplotlib/cbook.py#L2237-L2251

The pytorch tensor does not support any of the conversion methods, so Matplotlib doesn't really know what to do with it. There is a discussion in https://github.com/matplotlib/matplotlib/issues/22645 about this, but if I remember correctly we expect the libraries to support the `to_numpy` method (but still support the `values` attribute).

(I could not install jax, but I suppose something similar goes on there.)

**Comment by oscargus** [comment]
And when the conversion doesn't work, it ends up in this loop:
https://github.com/matplotlib/matplotlib/blob/b61bb0b6392c23d38cd45c658bfcd44df145830d/lib/matplotlib/cbook.py#L1332-L1348
which is where most of the time is spent.

**Comment by patel-zeel** [comment]
Thanks for the quick response, @oscargus! Given that both these libraries support `.__array__()` method for conversion to numpy array, wouldn't it be easier to add one more `if` condition in `_unpack_to_numpy` to include them?
```py
type(jax_array.__array__()), type(torch_array.__array__())
# Output: (numpy.ndarray, numpy.ndarray)
```

**Comment by oscargus** [comment]
Yes, I also noted that. It probably can make sense.

(I think the reason why we do this somewhat carefully is for unit information to not get lost.)

Would you be interested in submitting a patch? I think that if this goes last in the conversion chain, it shouldn't break too many things... (A problem here is that we do not, yet, test for "all" types that possibly can be used and "works". There's been a discussion of having a special test suite for that, but it has not yet been implemented.)

**Comment by patel-zeel** [comment]
Even `tensorflow` supports `.__array__()` method. I guess these 3 libraries account for almost 99% of the machine learning codebase available online :) It'd be great if this conversion passes without braking many things!

Sure, I'd submit a patch. I guess I need to only change the `_unpack_to_numpy` to the following, right?

```py
def _unpack_to_numpy(x): 
     """Internal helper to extract data from e.g. pandas and xarray objects.""" 
     if isinstance(x, np.ndarray): 
         # If numpy, return directly 
         return x 
     if hasattr(x, 'to_numpy'): 
         # Assume that any to_numpy() method actually returns a numpy array 
         return x.to_numpy() 
    if hasattr(x, '__array__'):
         # Assume that any to __array__() method returns a numpy array (e.g. `tensorflow`, `jax` and `pytorch` arrays)
         return x.__array__()
     if hasattr(x, 'values'): 
         xtmp = x.values 
         # For example a dict has a 'values' attribute, but it is not a property 
         # so in this case we do not want to return a function 
         if isinstance(xtmp, np.ndarray): 
             return xtmp 
     return x 
```

**Comment by timhoffm** [comment]
Yes, but please verify that `__array__` actually returns a numpy array, like we do with `values` above.

**Comment by patel-zeel** [comment]
Thank you for the important suggestion, @timhoffm. Now, `__array__` method check works in theory for the cases I imagined but `np.float32` type objects get stuck into that check. When `__array__` method is called on `np.float32` type object, it gets converted to `ndarray` type and eventually this leads to an infinite recursion. 

A temporary fix I could figure out is to add two more `if` conditions to check if object is of type `np.floating` (includes all float types) and type `np.integer` (includes all integer types including `uint`). I can also include a boolean check. Will it be all or this already looks like unpythonic?

Two more directions to solve this issue could be these.
* Raise an issue to add `to_numpy()` methods in JAX and PyTorch repos.
* Raise an issue to have a universal `numpy` object checker type in NumPy library so that we can replace `ndarray` check with that.

I am open to your suggestions.

**Comment by tacaswell** [comment]
I am not convinced this is a high-priority problem as there is a very simple user-side solution (cast to numpy before passing to Matplotlib!).

Given that, to my understanding, some of these arrays may actually be in GPU memory, forcing them back to CPU memory before passing to Matplotlib is probably a good idea anyway.

**Comment by jklymak** [comment]
I agrée this isn't high priority, but if an object has a numpy array representation, why would we ignore it?

**Comment by jklymak** [comment]
> To address the issue of the tensor being in GPU, a simple check on what device the tensor is on could be added. My execution crashed with 10K elements on CPU, which I expected to take less than a second to plot.

Just to set expectations, that is a request to torch, or something the user needs to take care of, not to Matplotlib.  We don't officially support torch or jax arrays, and we aren't going to have extra checks to support them beyond what `__array__` might support

**Comment by timhoffm** [comment]
Sorry this slipped.

The code suggested in *Edit1* of https://github.com/matplotlib/matplotlib/issues/25882#issuecomment-1546665040 looks reasonable to me.

Tests should include the numpy scalar case and a mock object with an `__array__` method (we don’t want to depend on torch or JAX).

@patel-zeel do you want to pick this up again.

**Comment by patel-zeel** [comment]
Sure, @timhoffm! Let me try this fix in my existing PR #25887.
